### DISCUSSION TITLE
STM32L496 SPI MOSI is wrong at clock 10mhz

### DISCUSSION THREAD

**Discussion by glin00** [zephyrproject-rtos/zephyr#38964] [Q&A] [answered]
Hello,
![SPI_CLK_10mhz_0x55](https://user-images.githubusercontent.com/20088622/135243544-f2a69e5b-4bbf-4341-b16a-f77ced902792.jpg)
![SPI_CLK_10mhz_0xaa](https://user-images.githubusercontent.com/20088622/135243552-c16dfdb7-5608-4416-beb3-142ad9987008.jpg)
![SPI_CLK_5mhz](https://user-images.githubusercontent.com/20088622/135243554-af12ec09-7949-49bb-89bd-87bede98e3d8.jpg)

Description: 
I am using ST MCU SPI-1 i/f to connect to OLED display.  When I configure Zephyr SPI clock to 5mhz, the OLED displays correctly. But when I configure Zephyr SPI clock to 10mhz, the OLED displays incorrectly. After I use an osilloscope to measure SPI mosi and clock pin, I found SPI clock has the right frequency 10mhz, but MOSI is wrong.

Hardware: ST MCU STM32L496AGI6P, OLED with SPI interface
Zephyr Ver: Zephyr OS  2.7.0-rc2   .   
OS: Linux Ubuntu

Analysis:
Attached SPI_CLK_5mhz.jpg is the 5mhz SPI clock when I try to write 0x55 0xaa. The OLED uses SPI clock rising edge to latch MOSI data.(CH1:SPI clock, CH2:SPI MOSI)
Attached SPI_CLK_10mhz_0x55.jpg  is the 10mhz SPI clock when I try to write 0x55.  
Attached SPI_CLK_10mhz_0xAA.jpg  is the10mhz SPI clock when I try to write 0xAA.  

When I build the firmware without Zephyr OS, using the same hardware as mentioned above , the MOSI is correct when SPI clock is configured at 20mhz. So the hardware should be no problem.

Below is how SPI and clock system get configured in DTS.

&clk_hsi {
	status = "okay";
};

&pll {
	div-m = <1>;
	mul-n = <20>;
	div-p = <7>;
	div-q = <2>;
	div-r = <4>;
	clocks = <&clk_hsi>;
	status = "okay";
};

&rcc {
	clocks = <&pll>;
	clock-frequency = <DT_FREQ_M(80)>;
	ahb-prescaler = <1>;
	apb1-prescaler = <1>;
	apb2-prescaler = <1>;
};

&spi1 {
	pinctrl-0 = <&spi1_sck_pa5 &spi1_miso_pb4 &spi1_mosi_pb5>;
	status = "okay";
};

Please advise me how to fix this problem. Thank You!

**Comment by erwango** [comment, marked as answer]
@glin00 can you try to change pin speed, adding the following to board.dts file
```
&spi1_mosi_pb5{
slew-rate = "very-high-speed";
};
```

**Reply by glin00** [reply]
@erwango with your advise, the MOSI is correct. Thanks.
